Unable to link regex crate with "-Z instrument-coverage" on x86_64-pc-windows-msvc

Hello,

Linking [regex](https://docs.rs/regex/1.5.4/regex/) crate on `x86_64-pc-windows-msvc` fails when using `-Z instrument-coverage`.
The link succeeds when using `-Z instrument-coverage=except-unused-functions`

```toml
# Cargo.toml
[package]
name = "foobar"
version = "0.1.0"
authors = ["foobar"]
edition = "2018"

[dependencies]
regex = "1.5.4"
```

```rust
// main.rs
fn main() {
    use regex::Regex;
    let re = Regex::new(r"^\d{4}-\d{2}-\d{2}$").unwrap();
    assert!(re.is_match("2014-01-01"));
}
```

I expected to see this happen: linking successful.

Instead, this happened:

```sh
error: linking with `link.exe` failed: exit code: 1227

note: libmemchr-40eebd4a74e50761.rlib(memchr-40eebd4a74e50761.memchr.2b5ledse-cgu.14.rcgu.o) : fatal error LNK1227: conflicting weak extern definition for '_RINvNtNtNtCs838Isxeq56n_6memchr6memmem9prefilter11genericsimd14find_in_chunk3pEB8_'.  New default '.weak._RINvNtNtNtCs838Isxeq56n_6memchr6memmem9prefilter11genericsimd14find_in_chunk3pEB8_.default._RINvMNtCs93I3MPx9COt_4core5sliceSh13get_uncheckedINtNtNtB5_3ops5range7RangeTojEECs838Isxeq56n_6memchr' conflicts with old default '.weak._RINvNtNtNtCs838Isxeq56n_6memchr6memmem9prefilter11genericsimd14find_in_chunk3pEB8_.default._RINvMNtNtCs93I3MPx9COt_4core3ptr9const_ptrPh4castuECs838Isxeq56n_6memchr' in libmemchr-40eebd4a74e50761.rlib(memchr-40eebd4a74e50761.memchr.2b5ledse-cgu.3.rcgu.o)
```

### Meta

`rustc --version --verbose`:
```
rustc +nightly --version --verbose       
rustc 1.54.0-nightly (3e99439f4 2021-05-17)
binary: rustc
commit-hash: 3e99439f4dacc8ba0d2ca48d221694362d587927
commit-date: 2021-05-17
host: x86_64-pc-windows-msvc
release: 1.54.0-nightly
LLVM version: 12.0.1
```

I'm still looking into this but I have managed to construct a minimal (I believe) repro:

```toml
# Cargo.toml
[package]
name = "memchr"
version = "2.4.1"
edition = "2018"
```

```rust
// src/main.rs
use memchr::memchr;

fn main() {
    memchr();
}
```

```rust
// src/lib.rs
mod cow {
  #[derive(Debug)]
  struct CowBytes<'a>(&'a ());

  #[derive(Debug)]
  enum Imp<'a> {
      Owned(&'a [u8]),
  }
}

pub fn memchr() {
  static FN: std::sync::Once = std::sync::Once::new();

  let _f = &FN.is_completed();
}

mod memmem;
```

```rust
mod genericsimd {
    use crate::memmem::PrefilterFnTy;
    
    #[inline(always)]
    pub(crate) fn find(
        haystack: &(),
        fallback: PrefilterFnTy,
    ) { }
    
    fn find_in_chunk2() { }    
}

mod x86 {
    pub(crate) mod avx {
        use core::arch::x86_64::__m256i;
    
        use crate::memmem::PrefilterFnTy;
        
        const _: PrefilterFnTy = find;
        
        pub(crate) fn find(
            haystack: &(),
        ) {
            super::super::genericsimd::find(
                haystack,
                super::sse::find2,
            )
        }
    }
    
    pub(crate) mod sse {
        use core::arch::x86_64::__m128i;

        pub(crate) fn find2(
            haystack: &(),
        ) {
            super::super::genericsimd::find(
                haystack,
                |_| { },
            )
        }   
    }    
}

#[derive(Debug)]
struct NeedleInfo { }

type PrefilterFnTy = fn(
    haystack: &(),
);
```

Which, with the appropriate flags, will then trigger the link error:

> \> $env::RUSTFLAGS="-Zinstrument-coverage -Ccodegen-units=16"
> \> cargo +nightly build
>    Compiling memchr v2.4.1 (D:\code\temp\memchr)
> error: linking with `link.exe` failed: exit code: 1227
>   |
>   = note: libmemchr-c0a7af6bade6d9a3.rlib(memchr-c0a7af6bade6d9a3.58nah2n3aa285jtv.rcgu.o) : fatal error LNK1227: conflicting weak extern definition for '_RNvNtNtCsgYqfNtRWHVc_6memchr6memmem11genericsimd14find_in_chunk2'.  New default '.weak._RNvNtNtCsgYqfNtRWHVc_6memchr6memmem11genericsimd14find_in_chunk2.default._RINvNtCsjJUIXhzXbX_4core3ptr13drop_in_placeRRShECsgYqfNtRWHVc_6memchr' conflicts with old default '.weak._RNvNtNtCsgYqfNtRWHVc_6memchr6memmem11genericsimd14find_in_chunk2.default._RNvCsgYqfNtRWHVc_6memchr6memchr' in libmemchr-c0a7af6bade6d9a3.rlib(memchr-c0a7af6bade6d9a3.4b1up16ivtuwqgl4.rcgu.o)

There's quite a lot of dead code in the "minimal" repro but removing any of it or even moving it around causes the error to go away. We need basically this exact structure to get the codegen units to be placed into the state which triggers this bug.

Ok, that was super helpful and pointed me in the direction of the actual minimal repro. Thanks @richkadel!

```rust
// lib.rs
mod foo {
    #[inline(always)]
    pub fn called() { }

    fn uncalled() { }
}

pub mod bar {
    pub fn call_me() {
        super::foo::called();
    }
}

pub mod baz {
    pub fn call_me() {
        super::foo::called();
    }
}
```

```rust
// repro.rs
use lib::{bar, baz};

fn main() {
    bar::call_me();
    baz::call_me();
}
```

```powershell
> rustc +nightly --crate-type lib -Zinstrument-coverage --crate-name lib lib.rs
> rustc +nightly --edition 2021 --extern lib=liblib.rlib repro.rs
error: linking with `link.exe` failed: exit code: 1227
  |
  = note: liblib.rlib(lib.lib.f2c86790-cgu.0.rcgu.o) : fatal error LNK1227: conflicting weak extern definition for '_RNvNtCslAVHmXkNn7E_3lib3foo8uncalled'.  New default '.weak._RNvNtCslAVHmXkNn7E_3lib3foo8uncalled.default._RNvNtCslAVHmXkNn7E_3lib3baz7call_me' conflicts with old default '.weak._RNvNtCslAVHmXkNn7E_3lib3foo8uncalled.default._RNvNtCslAVHmXkNn7E_3lib3bar7call_me' in liblib.rlib(lib.lib.f2c86790-cgu.1.rcgu.o)
```

Since `called` is marked `#[inline(always)]`, it's included in the CGU for `lib::bar` **and** the CGU for `lib::baz`. Therefore, both CGUs think they should be responsible for generating the unused function and there are two copies of it generated. While the functions are generated as weak symbols, the MSVC linker seems to treat weak symbols differently than unix-y linkers do. 

I'm a bit curious why we use `weak/hidden` instead of `private/default` for these functions? 

https://github.com/rust-lang/rust/blob/b27661eb33c74cb514dba059b47d86b6582ac1c2/compiler/rustc_codegen_llvm/src/coverageinfo/mod.rs#L215-L216

Using `private/default` seems to produce the correct results (at least in this simple test):

```
PS D:\code\temp> D:\code\rust4\build\x86_64-pc-windows-msvc\llvm\bin\llvm-cov.exe show -Xdemangler=rustfilt .\repro.exe -instr-profile="default.profdata"
lib.rs:
    1|       |mod foo {
    2|       |    #[inline(always)]
    3|      2|    pub fn called() { }
    4|       |
    5|      0|    fn uncalled() { }
    6|       |}
    7|       |
    8|       |pub mod bar {
    9|      1|    pub fn call_me() {
   10|      1|        super::foo::called();
   11|      1|    }
   12|       |}
   13|       |
   14|       |pub mod baz {
   15|      1|    pub fn call_me() {
   16|      1|        super::foo::called();
   17|      1|    }
   18|       |}
   19|       |

repro.rs:
    1|       |use lib::{bar, baz};
    2|       |
    3|      1|fn main() {
    4|      1|    bar::call_me();
    5|      1|    baz::call_me();
    6|      1|}
```

Changing to use those attributes instead seems like a quick, easy fix to resolve this for now. A better solution might be to reserve one CGU and put all of the dead code in that CGU which would sidestep the entire issue by ensuring there is only one CGU with these functions in it.

--- a/compiler/rustc_codegen_llvm/src/coverageinfo/mod.rs
+++ b/compiler/rustc_codegen_llvm/src/coverageinfo/mod.rs
@@ -212,8 +212,8 @@ fn declare_unused_fn(cx: &CodegenCx<'ll, 'tcx>, def_id: &DefId) -> Instance<'tcx
         ),
     );
 
-    llvm::set_linkage(llfn, llvm::Linkage::WeakAnyLinkage);
-    llvm::set_visibility(llfn, llvm::Visibility::Hidden);
+    llvm::set_linkage(llfn, llvm::Linkage::PrivateLinkage);
+    llvm::set_visibility(llfn, llvm::Visibility::Default);
 
     assert!(cx.instances.borrow_mut().insert(instance, llfn).is_none());
 

--- a/src/test/run-make-fulldeps/coverage-reports/expected_show_coverage.issue-85461.txt
+++ b/src/test/run-make-fulldeps/coverage-reports/expected_show_coverage.issue-85461.txt
@@ -1,0 +1,36 @@
+../coverage/issue-85461.rs:
+    1|       |// Regression test for #85461: MSVC sometimes fail to link with dead code and #[inline(always)]
+    2|       |
+    3|       |extern crate inline_always_with_dead_code;
+    4|       |
+    5|       |use inline_always_with_dead_code::{bar, baz};
+    6|       |
+    7|      1|fn main() {
+    8|      1|    bar::call_me();
+    9|      1|    baz::call_me();
+   10|      1|}
+
+../coverage/lib/inline_always_with_dead_code.rs:
+    1|       |// compile-flags: -Zinstrument-coverage -Ccodegen-units=4 -Copt-level=0
+    2|       |
+    3|       |#![allow(dead_code)]
+    4|       |
+    5|       |mod foo {
+    6|       |    #[inline(always)]
+    7|      2|    pub fn called() { }
+    8|       |
+    9|      0|    fn uncalled() { }
+   10|       |}
+   11|       |
+   12|       |pub mod bar {
+   13|      1|    pub fn call_me() {
+   14|      1|        super::foo::called();
+   15|      1|    }
+   16|       |}
+   17|       |
+   18|       |pub mod baz {
+   19|      1|    pub fn call_me() {
+   20|      1|        super::foo::called();
+   21|      1|    }
+   22|       |}
+

--- a/src/test/run-make-fulldeps/coverage/issue-85461.rs
+++ b/src/test/run-make-fulldeps/coverage/issue-85461.rs
@@ -1,0 +1,10 @@
+// Regression test for #85461: MSVC sometimes fail to link with dead code and #[inline(always)]
+
+extern crate inline_always_with_dead_code;
+
+use inline_always_with_dead_code::{bar, baz};
+
+fn main() {
+    bar::call_me();
+    baz::call_me();
+}

--- a/src/test/run-make-fulldeps/coverage/lib/inline_always_with_dead_code.rs
+++ b/src/test/run-make-fulldeps/coverage/lib/inline_always_with_dead_code.rs
@@ -1,0 +1,22 @@
+// compile-flags: -Zinstrument-coverage -Ccodegen-units=4 -Copt-level=0
+
+#![allow(dead_code)]
+
+mod foo {
+    #[inline(always)]
+    pub fn called() { }
+
+    fn uncalled() { }
+}
+
+pub mod bar {
+    pub fn call_me() {
+        super::foo::called();
+    }
+}
+
+pub mod baz {
+    pub fn call_me() {
+        super::foo::called();
+    }
+}

--- a/src/test/ui/issues/issue-85461.rs
+++ b/src/test/ui/issues/issue-85461.rs
@@ -1,0 +1,27 @@
+// compile-flags: -Zinstrument-coverage -Ccodegen-units=4 --crate-type dylib -Copt-level=0
+// build-pass
+// needs-profiler-support
+
+// Regression test for #85461 where MSVC sometimes fails to link instrument-coverage binaries
+// with dead code and #[inline(always)].
+
+#![allow(dead_code)]
+
+mod foo {
+    #[inline(always)]
+    pub fn called() { }
+
+    fn uncalled() { }
+}
+
+pub mod bar {
+    pub fn call_me() {
+        super::foo::called();
+    }
+}
+
+pub mod baz {
+    pub fn call_me() {
+        super::foo::called();
+    }
+}

